Lb/insert block button

luisbettencourt opened the pull request:
## 🌟 What is the purpose of this PR?

This PR adds a block insertion button between blocks.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202500568584719/f) _(internal)_
- [Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-App?node-id=3635%3A177743)

## 🔍 What does this change?

- Adds a button after each block;
- Clicking on it opens a block list (reusing `BlockSuggester`);
- Selecting a block on this list inserts a it below the block the button belongs to.

## 📹 Demo

![demo](https://user-images.githubusercontent.com/37453800/182439483-be666a47-9bb8-4f15-837f-9905358ef3d5.gif)

luisbettencourt commented:
Should we also add the button above the first block, to insert a block at the start of the document? @nonparibus

nonparibus commented:
> Should we also add the button above the first block, to insert a block at the start of the document? @nonparibus 

Please!

luisbettencourt commented:
Done in a50edc467ae0aca006b750221f33721e7475fb69

teenoh commented on packages/hash/frontend/src/shared/icons/plus-box-outline-icon.tsx:
```diff
@@ -0,0 +1,13 @@
+    <SvgIcon {...props} width="24" height="24" viewBox="0 0 24 24" fill="none">
+      <path
+        d="M17.25 5.25C18.078 5.25 18.75 5.922 18.75 6.75V17.25C18.75 18.0787 18.078 18.75 17.25 18.75H6.75C5.922 18.75 5.25 18.0787 5.25 17.25V6.75C5.25 5.922 5.922 5.25 6.75 5.25H17.25ZM17.2501 17.25V6.75H6.75007V17.2507L17.2501 17.25ZM12.7506 8.25143H11.2506V11.2514H8.2506V12.7514H11.2506V15.7514H12.7506V12.7514H15.7506V11.2514H12.7506V8.25143Z"
+        fill="#C1CFDE"
```
Let's replace this with `currentcolor`, so it's easier to reuse and change the color of the icon

luisbettencourt replied:
Done in a2d7864e2a1ea6011b8414a67dc145bc20b9d5db

teenoh commented on packages/hash/frontend/src/blocks/page/InsertBlock.tsx:
```diff
@@ -0,0 +1,85 @@
+export const InsertBlock: FunctionComponent<InsertBlockProps> = ({
+  onBlockSuggesterChange,
+}) => {
+  const [contextMenu, setContextMenu] = useState<{
```
Nit: I think we should call this contextMenuPosition for more clarity

luisbettencourt replied:
Done in 7f45bcd0b2619a71d3e5d1c378902803c1cddf97

teenoh commented on packages/hash/frontend/src/blocks/page/style.module.css:
```diff
@@ -63,3 +64,19 @@
   overflow-x: auto;
 }
+
+.Block__InsertBlock {
+  width: 100%;
+  max-width: 1200px;
+  position: absolute;
+  height: 30px;
+  margin-left: 32px;
+}
+
+.Block__InsertBlock__Bottom {
+  top: 100%;
+}
+
+.Block__InsertBlock__Top {
+  bottom: 100%;
+}
```
Nit: I'd prefer if we have these styles in InsertBlock. I think it makes sense to use this file for scenarios where we can't use MUI to style the element in question. It's not a blocking change though and if it seems difficult, then we can leave as is

luisbettencourt replied:
I see your point but these are related to how we position the `InsertBlock` component inside the block view, which I think should be abstracted from component itself, and be defined in the component that renders it. We only use this component here, but would we reuse it somewhere else we probably wouldn't want this styles there.

teenoh reviewed (commented):
Thanks for working on this @luisbettencourt! I tested and it works well 🎉. I added some comments, other than those, it looks good to me

CiaranMn reviewed (changes requested):
Thanks @luisbettencourt - I spotted a few things it would be good to address:

1. The line should be aligned with the left side of the block content - it's currently sticking out to the left too much at some sizes
<img width="994" alt="Screenshot 2022-08-04 at 10 00 19" src="https://user-images.githubusercontent.com/37743469/182811788-eca0e5d7-1d78-41b2-b7ac-b502d55d6464.png">

2. If you drag a block to the first position, it does not gain an 'insert above' button

3. Inserting below the bottom block does not work

4. When you create a block, we should focus inside the block (as happens when you insert via the / command) – at the moment, the focus jumps back to the previously focused block, which can mean the screen jumps around if you had a paragraph a few blocks above focused

luisbettencourt commented:
> Thanks @luisbettencourt - I spotted a few things it would be good to address:
> 
> 1. The line should be aligned with the left side of the block content - it's currently sticking out to the left too much at some sizes
> 
> <img alt="Screenshot 2022-08-04 at 10 00 19" width="994" src="https://user-images.githubusercontent.com/37743469/182811788-eca0e5d7-1d78-41b2-b7ac-b502d55d6464.png">
> 
> 2. If you drag a block to the first position, it does not gain an 'insert above' button
> 3. Inserting below the bottom block does not work
> 4. When you create a block, we should focus inside the block (as happens when you insert via the / command) – at the moment, the focus jumps back to the previously focused block, which can mean the screen jumps around if you had a paragraph a few blocks above focused

Addressed in 0291f1b5a2249145e61e3b441192f122307a575e and 3c5fd5cff4f45f307e180e9269547ec596a90aa4.
Also fixed a bug where the document was scrolling to the top of the page whenever the block suggester menu was closed.

CiaranMn commented on packages/hash/frontend/src/blocks/page/BlockView.tsx:
```diff
@@ -218,7 +237,7 @@ export class BlockView implements NodeView<Schema> {
             );
 
-            this.editorView.dispatch(tr);
+            // this.editorView.dispatch(tr);
```
This is probably what has caused drag and drop to stop working
```suggestion
            this.editorView.dispatch(tr);
```

CiaranMn reviewed (changes requested):
Thanks @luisbettencourt, the bugs I mentioned have been addressed – dragging blocks no longer works though, I suspect because some code has been commented out

CiaranMn reviewed (approved):
Excellent, let's get it in!
